text_generation/causal_lm/cpp/continuous_batching/library/src/sequence_group.hpp

```diff
@@ -384,4 +410,54 @@ class SequenceGroup {
         }
         return false;
     }
+    
+    GenerationStream::Ptr get_generation_stream() {
+        return m_generation_stream;
+    }
+
+    void set_generation_status(GenerationStatus status) {
+        m_generation_stream->set_generation_status(status);
+    }
+
+    bool handle_dropped() {
+        return m_generation_stream->get_status() == GenerationStatus::DROPPED_BY_HANDLE;
+    }
+
+    void notify_handle() {
+        GenerationOutputs outputs;
+
+        // For beam search streaming is not available, so we notify only upon finishing
+        if(m_sampling_params.is_beam_search()) {
+            if (has_finished()) {
+                std::vector<Sequence::CPtr> finished_sequences = get_finished_sequences();
+
+                OPENVINO_ASSERT(finished_sequences.size() == num_total_seqs() && has_finished());
+                for (auto& sequence: finished_sequences) {
+                    GenerationOutput output;
+                    output.generated_token_ids = sequence->get_generated_ids();
+                    output.score = sequence->get_beam_search_score(m_sampling_params);
+                    outputs.emplace(sequence->get_grouped_id(), output);
+                }
+
+                if (outputs.size()) {
+                    m_generation_stream->push(outputs);
+                }
+            }
+        // For greedy or multinomial sampling we decide whever to stream partial results depending on the user parameter
+        } else if (m_sampling_params.is_greedy_sampling() || m_sampling_params.is_multinomial()) {
+            // TO DO: Now we always stream for greedy search for the sake of benchmarking
+            if (true /* m_sampling_params.stream */) {
+                for (auto& sequence : m_sequences) {
+                        outputs.emplace(sequence->get_grouped_id(), sequence->get_last_generation_output(m_sampling_params));
+                }
+                m_generation_stream->push(outputs);
+            }
+        }
+
+        if (out_of_memory()) {
+            set_generation_status(GenerationStatus::IGNORED);
+        } else if (has_finished()) {
+            set_generation_status(GenerationStatus::FINISHED);
```

Comment:
should we set status before actual cond. variable notification? As I see we can push new tokens, which triggers client, then we can set `FINISHED` status. As a client I want to be notified about all actions at the same time.

Reply:
The idea was that changing the status is part of interaction with the handle and all interactions with the handle were supposed to be in `notify_handle` method. Effectively we are doing just that - first we push the tokens, then we set `FINISHED` status. If by "the same time" you mean that you don't want to separately check status and read tokens then we would need to pack the status into the outputs we push. Is that what you mean?